File: Set 7/Course Scheduler/Scheduler.cpp
#include <iostream>
#include <vector>
using namespace std;

class Pair {
public:
    int first, second;
    Pair(int f, int s) : first(f), second(s) {}
};

vector<vector<int>> define(int tasks, const vector<Pair>& pre) {
    vector<vector<int>> graph(tasks);
    for (const auto& p : pre) {
        graph[p.second].push_back(p.first);
    }
    return graph;
}

bool dfs(const vector<vector<int>>& graph, int node, vector<bool>& onpath, vector<bool>& visited) {
    if (visited[node]) {
        return false;
    }
    onpath[node] = visited[node] = true;
    for (int neigh : graph[node]) {
        if (onpath[neigh] || dfs(graph, neigh, onpath, visited)) {
            return true;
        }
    }
    return false;
}

bool finish(int tasks, const vector<Pair>& pre) {
    vector<vector<int>> graph = define(tasks, pre);
    vector<bool> onpath(tasks, false);
    vector<bool> visited(tasks, false);
    for (int i = 0; i < tasks; ++i) {
        if (!visited[i] && dfs(graph, i, onpath, visited)) {
            return false;
        }
    }
    return true;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int numTasks, numPrerequisites;
        cin >> numTasks >> numPrerequisites;
        vector<Pair> prerequisites;
        for (int i = 0; i < numPrerequisites; ++i) {
            int first, second;
            cin >> first >> second;
            prerequisites.push_back(Pair(first, second));
        }
        
        if (finish(numTasks, prerequisites)) {
            cout << "YES" << endl;
        } else {
            cout << "NO" << endl;
        }
    }
    return 0;
}

File: Set 7/Course Scheduler/Scheduler.py
class pair:
    def __init__(self, first, second):
        self.first = first
        self.second = second

def define(tasks, pre):
    graph = []
    for i in range(tasks):
        graph.append([])

    for p in pre:
        graph[p.second].append(p.first)

    return graph

def dfs(graph, node, onpath, visited):
    if visited[node]:
        return False
    onpath[node] = visited[node] = True
    for neigh in graph[node]:
        if (onpath[neigh] or dfs(graph, neigh, onpath, visited)):
            return True
    return False

def finish(task, pre):
    graph = define(task, pre)
    onpath = [False] * task
    visited = [False] * task
    for i in range(task):
        if (not visited[i] and dfs(graph, i, onpath, visited)):
            return False
    return True

t = int(input())
for _ in range(t):
    numTasks = int(input())
    numPrerequisites = int(input())
    prerequisites = []
    for _ in range(numPrerequisites):
        first, second = map(int, input().split())
        prerequisites.append(pair(first, second))
    
    if finish(numTasks, prerequisites):
        print("YES")
    else:
        print("NO")
